### PR TITLE
HPCC-32072 Thor worker cleaner shutdown

### PR DESCRIPTION
**Pull request by mckellyln**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by mckellyln** [comment]
Not sure if this is really necessary but do like when Thor shuts down clean (esp when trying to debug MODULE_EXIT code).

File: thorlcr/slave/thslavemain.cpp [hunk 1]
```diff
@@ -285,19 +283,58 @@ bool UnregisterSelf(IException *e)
...
+
 bool ControlHandler(ahType type)
 {
+    static bool recvdSig = false;
+    if (recvdSig)
+    {
+        if (ahInterrupt == type)
+            _exit(128+SIGINT);
+        else
+            _exit(128+SIGTERM);
```
Comment (jakesmith) [inline thorlcr/slave/thslavemain.cpp:319]: is there any significance to the 128+ ?

File: thorlcr/slave/thslavemain.cpp [hunk 2]
```diff
@@ -285,19 +283,58 @@ bool UnregisterSelf(IException *e)
     return false;
 }
 
+void signalInFuture(int signo, unsigned timeout)
```
Comment (jakesmith) [inline thorlcr/slave/thslavemain.cpp:286]: I know it's seconds, but it's not clear from the prototype/parameter names, perhaps rename param to timeoutSecs, or method to signalInSecs

Comment (jakesmith) [reply thorlcr/slave/thslavemain.cpp:286]: don't know where else this method might be useful, but probably belongs in somewhere like jexcept

File: system/mp/mpcomm.cpp [hunk 3]
```diff
@@ -2677,10 +2678,13 @@ CMPServer::CMPServer(unsigned __int64 _role, unsigned _port, bool _listen)
 CMPServer::~CMPServer()
 {
 #ifdef _TRACEORPHANS
-    StringBuffer buf;
-    getReceiveQueueDetails(buf);
-    if (buf.length())
-        LOG(MCdebugInfo, "MP: Orphan check\n%s",buf.str());
+    if (dumpQueue)
```
Comment (jakesmith) [inline system/mp/mpcomm.cpp:2681]: is this really an incidental change? Why suppress dumping of unconsumed MP messages if slave has been told to shutdown by manager, vs if it hasn't?

**Review by jakesmith — COMMENTED** [review]
@mckellyln - a few questions.

File: thorlcr/slave/thslavemain.cpp (hunk 1 above)
Comment (mckellyln) [reply thorlcr/slave/thslavemain.cpp:319]: Standard shell exit values are 128 + signal

File: system/mp/mpcomm.cpp (hunk 3 above)
Comment (mckellyln) [reply system/mp/mpcomm.cpp:2681]: On clean shutdown I didn't want the MP Orphan check dump in the logs

File: thorlcr/slave/thslavemain.cpp (hunk 2 above)
Comment (mckellyln) [reply thorlcr/slave/thslavemain.cpp:286]: I will change timeout to timeoutsec.
I can move this into jlib if you want.

Comment (mckellyln) [reply thorlcr/slave/thslavemain.cpp]: changed to timeoutSec

Comment (mckellyln) [reply thorlcr/slave/thslavemain.cpp]: I can move signalInFuture() into jexcept.cpp if you think its better there.

File: system/mp/mpcomm.cpp (hunk 3 above)
Comment (jakesmith) [reply system/mp/mpcomm.cpp:2686]: ok, I don't think message are useful in general.

File: thorlcr/slave/thslavemain.cpp (hunk 2 above)
Comment (jakesmith) [reply thorlcr/slave/thslavemain.cpp]: not a big deal, but yes think it better to group the code there.

Comment (mckellyln) [reply thorlcr/slave/thslavemain.cpp]: ok, moved into jexcept.cpp

**Comment by mckellyln** [comment]
@jakesmith one more time, I've also rebased onto candidate-9.8.x.

**Review by jakesmith — APPROVED** [review]
@mckellyln - looks good. Please squash.

**Comment by mckellyln** [comment]
Squashed

**Pull request merged**